src/lib/Types/Primitives.hs

```diff
@@ -263,6 +263,8 @@ instance Store PtrLitVal where
 
 data LitVal = Int64Lit   Int64
             | Int32Lit   Int32
+            | Nat64Lit   Word64
+            | Nat32Lit   Word32
```

Comment:
I don't understand why are we extending the set of base types with `Nat`. To me, `NatXY` is just a newtype wrapper around `WordXY` that forgot how to do subtraction (`WordXY` are groups with well defined wrap-around, which we don't want from natural numbers). As such, the newtype should be defined in user space as
```
data Nat32 = MkNat32 (Word32)
```
and then, they should get unpacked before they're used in index-related primitives, which would not be parameterized by `Word`s:
```
def Fin ((MkNat n):Nat) = %Fin n
```

You could even keep using `IntXY` as the implementation of `Nat` and it would work too, without any changes to the compiler.